frontend/src/components/AddModal.js

```diff
@@ -0,0 +1,234 @@
+import React, { useState } from "react";
+import {
+  Button,
+  Modal,
+  ModalHeader,
+  ModalBody,
+  Row,
+  Col,
+  Label,
+} from "reactstrap";
+import { Control, LocalForm, Errors } from "react-redux-form";
+
+const required = (val) => val && val.length;
+const maxLength = (len) => (val) => !val || val.length <= len;
+const minLength = (len) => (val) => val && val.length >= len;
+const isNumber = (val) => !isNaN(Number(val));
+const validEmail = (val) =>
+  /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(val);
+
+function AddModal(props) {
+  const [modal, setModal] = useState(false);
+
+  const toggle = () => {
+    setModal(!modal);
+  };
+
+  const handleSubmit = (values) => {
+    toggle();
+    alert(JSON.stringify(values));
+    props.addUser(values);
+  };
+
+  return (
+    <>
+      <Button outline onClick={toggle}>
+        Add User
+      </Button>
+      <Modal isOpen={modal} toggle={toggle}>
+        <ModalHeader toggle={toggle}>Edit User</ModalHeader>
+        <ModalBody>
+          <LocalForm onSubmit={(values) => handleSubmit(values)}>
+            <Row className="form-group">
+              <Label htmlFor="username" md={2}>
+                Username
+              </Label>
+              <Col md={10}>
+                <Control.text
+                  model=".username"
+                  id="username"
+                  name="username"
+                  placeholder="Enter Username"
+                  className="form-control"
+                  validators={{
+                    required,
+                    minLength: minLength(3),
+                    maxLength: maxLength(20),
+                  }}
+                />
+                <Errors
+                  className="text-danger"
+                  model=".username"
+                  show="touched"
+                  messages={{
+                    required: "Required ",
+                    minLength: "Must be greater than 2 characters ",
+                    maxLength: "Must be 20 characters or less",
+                  }}
+                />
+              </Col>
+            </Row>
+
+            <Row className="form-group">
+              <Label htmlFor="name" md={2}>
+                Name
+              </Label>
+              <Col md={10}>
+                <Control.text
+                  model=".name"
+                  id="name"
+                  name="name"
+                  placeholder="Enter Name"
+                  className="form-control"
+                  validators={{
+                    required,
+                    minLength: minLength(3),
+                    maxLength: maxLength(20),
+                  }}
+                />
+                <Errors
+                  className="text-danger"
+                  model=".username"
```

Comment:
is this same error model or different?

Reply:
Redux error management for forms.
Every form component will have their own error types. 
Some components have the same types.
The errors are declared at the top.